File: src/models/Post.js
const mongoose = require('mongoose');

const commentSchema = new mongoose.Schema({
  authorId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  authorName: {
    type: String,
    required: true,
  },
  text: {
    type: String,
    required: true,
    trim: true,
  },
  createdAt: {
    type: Date,
    default: Date.now,
  },
}, {
  timestamps: true,
});

const postSchema = new mongoose.Schema({
  authorId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  authorName: {
    type: String,
    required: true,
  },
  authorAddress: {
    type: String,
    required: true,
  },
  text: {
    type: String,
    required: true,
    trim: true,
  },
  imagePath: {
    type: String,
  },
  location: {
    type: {
      type: String,
      enum: ['Point'],
      default: 'Point',
    },
    coordinates: {
      type: [Number],
      required: true,
    },
  },
  likedBy: [{
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
  }],
  comments: [commentSchema],
  createdAt: {
    type: Date,
    default: Date.now,
  },
  updatedAt: {
    type: Date,
    default: Date.now,
  },
  isDeleted: {
    type: Boolean,
    default: false,
  },
}, {
  timestamps: true,
});

// Индекс для геопространственных запросов
postSchema.index({ location: '2dsphere' });
postSchema.index({ createdAt: -1 });

// Виртуальные поля
postSchema.virtual('likesCount').get(function() {
  return this.likedBy.length;
});

postSchema.virtual('commentsCount').get(function() {
  return this.comments.length;
});

// Метод для поиска постов в радиусе
postSchema.statics.findNearby = function(coordinates, maxDistance = 500, limit = 20) {
  return this.find({
    location: {
      $near: {
        $geometry: {
          type: 'Point',
          coordinates: coordinates,
        },
        $maxDistance: maxDistance,
      },
    },
    isDeleted: false,
  })
  .sort({ createdAt: -1 })
  .limit(limit)
  .populate('authorId', 'firstName lastName displayName avatarUrl');
};

module.exports = mongoose.model('Post', postSchema); 

File: lib/models/post.dart
class Post {
  final String id;
  final String authorName;
  final String authorAddress;
  final String text;
  final String? imagePath;
  final DateTime createdAt;
  final List<String> likedBy;
  final List<Comment> comments;

  Post({
    required this.id,
    required this.authorName,
    required this.authorAddress,
    required this.text,
    this.imagePath,
    required this.createdAt,
    List<String>? likedBy,
    List<Comment>? comments,
  }) : likedBy = likedBy ?? [],
       comments = comments ?? [];

  int get likesCount => likedBy.length;
  int get commentsCount => comments.length;

  Post copyWith({
    String? id,
    String? authorName,
    String? authorAddress,
    String? text,
    String? imagePath,
    DateTime? createdAt,
    List<String>? likedBy,
    List<Comment>? comments,
  }) {
    return Post(
      id: id ?? this.id,
      authorName: authorName ?? this.authorName,
      authorAddress: authorAddress ?? this.authorAddress,
      text: text ?? this.text,
      imagePath: imagePath ?? this.imagePath,
      createdAt: createdAt ?? this.createdAt,
      likedBy: likedBy ?? this.likedBy,
      comments: comments ?? this.comments,
    );
  }
}

class Comment {
  final String id;
  final String authorName;
  final String text;
  final DateTime createdAt;

  Comment({
    required this.id,
    required this.authorName,
    required this.text,
    required this.createdAt,
  });
} 